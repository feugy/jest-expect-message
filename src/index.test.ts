import 'jest-extended';
import './index';

describe('jest-expect-message', () => {
  test('should fail with custom message', () => {
    expect(() =>
      expect(false, 'Woah this should be false!').toBeTruthy()
    ).toThrowErrorMatchingSnapshot();
  });

  test('should fail without custom message', () => {
    expect(() => expect(false).toBeTruthy()).toThrowErrorMatchingSnapshot();
  });

  test('should pass when given custom message', () => {
    expect(1, 'should have been 1').toBe(1);
  });

  test('should pass when not given custom message', () => {
    expect(1).toBe(1);
  });

  describe('supports custom matchers registered after jest-expect-message', () => {
    expect.extend({
      toBeDivisibleBy(received, argument) {
        const pass = received % argument == 0;
        const message = pass
          ? () => `expected ${received} not to be divisible by ${argument}`
          : () => `expected ${received} to be divisible by ${argument}`;
        return { message, pass };
      },
    });

    interface ExtendedAsymmetricMatchers
      extends jest.InverseAsymmetricMatchers {
      toBeDivisibleBy(num: number): void;
    }

    interface ExtendedJestMatchers<R> extends jest.Matchers<R> {
      toBeDivisibleBy(num: number): void;
      not: ExtendedAsymmetricMatchers;
    }

    interface ExtendedExpect extends jest.Expect {
      <T = any>(actual: T, customMessage?: string): ExtendedJestMatchers<T>;
      toBeDivisibleBy(num: number): void;
      not: ExtendedAsymmetricMatchers;
    }

    const extendedExpect = expect as ExtendedExpect;

    it('allows new custom matchers to use custom messages', () => {
      expect(() =>
        extendedExpect(100, '100 % 3 === 1').toBeDivisibleBy(3)
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        extendedExpect(101, '101 % 1 === 0').not.toBeDivisibleBy(1)
      ).toThrowErrorMatchingSnapshot();
    });

    it('supports custom asymmetric matchers', () => {
      expect({ apples: 6, bananas: 3 }).toEqual({
        apples: extendedExpect.toBeDivisibleBy(1),
        bananas: extendedExpect.not.toBeDivisibleBy(2),
      });
    });
  });
});
