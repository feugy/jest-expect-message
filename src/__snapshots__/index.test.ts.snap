// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-expect-message should fail with custom message 1`] = `"matcherResult.message is not a function"`;

exports[`jest-expect-message should fail without custom message 1`] = `"result.message is not a function"`;

exports[`jest-expect-message supports custom matchers registered after jest-expect-message allows new custom matchers to use custom messages 1`] = `"matcherResult.message is not a function"`;

exports[`jest-expect-message supports custom matchers registered after jest-expect-message allows new custom matchers to use custom messages 2`] = `"matcherResult.message is not a function"`;
